File: programmers/javascript/level3/베스트앨범.js
const genres = ["classic", "pop", "classic", "classic", "pop"];
const plays = [500, 600, 150, 800, 2500];

const sum = (a) => {
  return a.reduce((acc, cur) => {
    return acc + cur[0];
  }, 0);
};

const solution = (genres, plays) => {
  const answer = [];
  const genre = [];
  const play = [];
  for (let i = 0; i < genres.length; i++) {
    if (genre.includes(genres[i])) {
      //console.log(genre);
      const idx = genre.indexOf(genres[i]);
      play[idx].push([plays[i], i]);
    } else {
      genre.push(genres[i]);
      play.push([[plays[i], i]]);
    }
  }

  play.sort((a, b) => {
    return sum(b) - sum(a);
  });

  for (let i = 0; i < play.length; i++) {
    play[i].sort((a, b) => {
      return b[0] - a[0];
    });
  }

  for (let i = 0; i < play.length; i++) {
    for (let j = 0; j < play[i].length; j++) {
      if (j === 2) break;
      answer.push(play[i][j][1]);
    }
  }
  return answer;
};

console.log(solution(genres, plays));

File: python/programmers/level3/베스트앨범.py
def solution(genres, plays):
    answer = []

    genreDict = {}
    genreArr = []

    for i, (genre, play) in enumerate(zip(genres, plays)):
        if genre not in genreDict:
            genreDict[genre] = []
        genreDict[genre].append([i, play])

    for g in genreDict:
        genreDict[g].sort(key=lambda x: x[1], reverse=True)
        genreArr.append([g, sum(play for _, play in genreDict[g])])

    genreArr.sort(key=lambda x: x[1], reverse=True)

    # print(genreDict)
    # print(genreArr)
    for g, _ in genreArr:
        answer.extend([x[0] for x in genreDict[g][:2]])
    return answer


print(solution(['classic', 'pop', 'classic', 'classic', 'pop'], [500, 600, 150, 800, 2500]))
